src/msg/producer/writer/message_writer.go

```diff
@@ -261,7 +262,10 @@ func (w *messageWriterImpl) write(
 		written = false
 	)
 	for i := len(iterationIndexes) - 1; i >= 0; i-- {
-		if err := consumerWriters[randIndex(iterationIndexes, i)].Write(w.encoder.Bytes()); err != nil {
+		// NB(r): Always select the same connection index per connection.
+		consumerWriter := consumerWriters[randIndex(iterationIndexes, i)]
+		connIndex := int(w.replicatedShardID % uint64(w.numConnections))
```

Comment:
This doesn't depend on the index; can do it outside the loop to avoid recalculating?